### DISCUSSION TITLE
Performance tips, when creating custom indicators

### DISCUSSION THREAD

**Discussion by mgamache** [DaveSkender/Stock.Indicators#1017] [Help and support]
I am in the process of optimizing strategies with 2-3 indicators on five minute charts (so there are a lot of bars each day). Are there any best practices for optimizing the performance of the indicators? I am limiting the data to a list with 350 items and the values seem to match Trading View so it's enough data.  

Indicators used (some as part of my own custom indicators): Stochastic VWMA, EMA, SMA. 

any tips would be helpful. 

thanks

**Comment by DaveSkender** [comment]
There’s quite a few aspects of your implementation that might have performance tuning opportunities.  If you don’t know the source of slowness, the first step is to test each unit, with something like [benchmarks.net](https://benchmarkdotnet.org/) or with other IDE profiler tools, so you can 1) see what part is slower, and 2) try different things to see how it affects performance.

**Reply by DaveSkender** [reply]
Lessons learned from my experience on this project, for your custom indicators:

- LINQ syntax is easier and nicer to code, but 5-10x slower
- beware of converting collection types or data types in unneeded ways
- non-nullable and lower precision floating point data types operate 2x faster

**Reply by DaveSkender** [reply]
I’d have to see more of your implementation approach to help more.

**Reply by mgamache** [reply]
Thanks, I did a quick CPU profile just to make sure the indicators are taking most of the CPU. I use doubles and not decimals for all indicator calculations. 

I've attached one of my custom indicators based on this TV version for your review. 
[https://www.tradingview.com/v/tqitSsyG/](https://www.tradingview.com/v/tqitSsyG/)

[BBWPIndicator.zip](https://github.com/DaveSkender/Stock.Indicators/files/10659681/BBWPIndicator.zip)

**Reply by DaveSkender** [reply]
These LINQ elements are really slowing this down.
`ElementAt`, `Take`, `Skip`, `Select`, `Sum`
There's also a faster inline way to do your initial SMA, and StdDev
Also, `enum` definition might be better outside of the result class.

**Reply by DaveSkender** [reply]
Checkout the *.Series.cs [files in our sources](https://github.com/DaveSkender/Stock.Indicators/tree/main/src) for examples

**Reply by mgamache** [reply]
Thanks the profiler didn't flag the Skip Select or Sum as major contributors, but the SMA/StdDev were large CPU percentages so that will be the first place I look. Then I'll look at replacing the Linq statements.

note: the Select statement was put in place as an optimization vs a for loop that was slower.

**Reply by DaveSkender** [reply]
Yeah, with the way you’re setup, you have three “passes” over your quotes, so it’ll be at least 3x slower than a single pass approach.  Both SMA and STDEV are particularly intensive because they have to re-evaluate the lookback window with every new bar.
